LoadKeyboardLayout(): Returns NULL when the library is loaded from the wrong directory

davidebeatrici opened the pull request:
`GetLastError()` returns 0 too.

Behavior encountered on Windows 10, 8.1 and 7. I didn't try on older versions.

---

CMake project: [Test.zip](https://github.com/MicrosoftDocs/sdk-api/files/6199791/Test.zip)
Registry import file: [a0000409.zip](https://github.com/MicrosoftDocs/sdk-api/files/6138520/a0000409.zip)

PRMerger6 commented:
@davidebeatrici : Thanks for your contribution! The author(s) have been notified to review your proposed change.

davidebeatrici commented:
Same behavior with `KBDUSINTLALTGR.DLL`.

davidebeatrici commented:
The code I'm using:
```c
#include <stdio.h>
#include <stdlib.h>

#ifndef WIN32_LEAN_AND_MEAN
#define WIN32_LEAN_AND_MEAN
#endif

#include <windows.h>

#define ILOT_INSTALL 0x00000000

typedef HRESULT (WINAPI *INSTALLLAYOUTORTIP)(LPCWSTR psz, DWORD dwFlasg);

int ShowMessage(const char *title, const char *message, const unsigned int type, const va_list args)
{
	char buf[1024];
	vsnprintf(buf, sizeof(buf), message, args);
	return MessageBox(NULL, buf, title, type);
}

int ShowInfo(const char *title, const char *message, ...)
{
	va_list args;
	va_start(args, message);
	const int ret = ShowMessage(title, message, MB_OK | MB_ICONINFORMATION, args);
	va_end(args);

	return ret;
}

int main()
{
	HKL hkl = LoadKeyboardLayout("a0000409", 0);
	ShowInfo("LoadKeyboardLayout()", "%p, %lu", hkl, GetLastError());

	HMODULE hDLL = LoadLibrary("input.dll");
	INSTALLLAYOUTORTIP InstallLayoutOrTip = (INSTALLLAYOUTORTIP)GetProcAddress(hDLL, "InstallLayoutOrTip");

	BOOL ret = (*InstallLayoutOrTip)(L"0x0409:0xa0000409", ILOT_INSTALL);
	ShowInfo("InstallLayoutOrTip()", "%u", ret);

	FreeLibrary(hDLL);

	const int size = GetKeyboardLayoutList(0, NULL);
	HKL *list = malloc(sizeof(HKL) * size);
	GetKeyboardLayoutList(size, list);

	for (int i = 0; i < size; ++i) {
		ShowInfo("HKL", "%p", list[i]);
	}

	free(list);

	char keyboard[KL_NAMELENGTH];
	GetKeyboardLayoutName(keyboard);
	ShowInfo("Current layout", "%s", keyboard);

	return 0;
}
```

Here's the registry import file: [a0000409.zip](https://github.com/MicrosoftDocs/sdk-api/files/6138520/a0000409.zip)

davidebeatrici commented:
I can consistently reproduce the issue with https://github.com/davidebeatrici/kli: `LoadKeyboardLayout()` always returns 0.

Interestingly, the function succeeds if I call it with the exact same arguments in another application.

`InstallLayoutOrTip()` calls the function internally:

![image](https://user-images.githubusercontent.com/5897523/112230227-d8e4e680-8c34-11eb-91ae-133743cb801d.png)

It probably ignores the return value of `LoadKeyboardLayout()` because it returns `TRUE` even if the layout is not loaded.

davidebeatrici committed:
LoadKeyboardLayout(): Returns NULL when the library is loaded from the wrong directory

GetLastError() returns 0 too.

Behavior encountered on Windows 10, 8.1 and 7. I didn't try on older versions.

PRMerger17 commented:
@davidebeatrici : Thanks for your contribution! The author(s) have been notified to review your proposed change.

davidebeatrici commented:
I submitted a report (**VULN-045212**) at https://msrc.microsoft.com/create-report before disclosing the real issue, to make sure that it was not considered serious.

I was informed that the report falls under "Application Directory (App Dir) DLL planting" in https://msrc-blog.microsoft.com/2018/04/04/triaging-a-dll-planting-vulnerability.

However, my proposed mitigation is still valid:
> I would specify the absolute path in the "Layout File" key for all layouts, for example: `%SystemRoot%\system32\KBDUS.DLL`
>
> This would also be consistent with "Layout Display Name" (e.g. `@%SystemRoot%\system32\input.dll,-5000`).

In addition to that, I would expect `NtUserLoadKeyboardLayoutEx()` to call `LoadLibraryEx()` with `LOAD_LIBRARY_AS_DATAFILE` (as suggested in https://msrc-blog.microsoft.com/2014/05/13/load-library-safely).

davidebeatrici commented:
Turns out "Layout File" cannot be `REG_EXPAND_SZ`: `LoadKeyboardLayout()` fails with `ERROR_MOD_NOT_FOUND`.

This happens because the string is not expanded:
![](https://user-images.githubusercontent.com/5897523/112389733-828bac80-8cf5-11eb-82d0-dda8c76e2162.png)

As a workaround, I'm simply going to remove the application directory from the DLL search paths in my project:
```c
SetDefaultDllDirectories(LOAD_LIBRARY_SEARCH_SYSTEM32);
```

Karl-Bridge-Microsoft committed:
Update nf-winuser-loadkeyboardlayouta.md

revised slightly

PRMerger13 commented:
@Karl-Bridge-Microsoft : Thanks for your contribution! The author(s) have been notified to review your proposed change.

Karl-Bridge-Microsoft committed:
Update nf-winuser-loadkeyboardlayoutw.md

revised slightly

PRMerger9 commented:
@Karl-Bridge-Microsoft : Thanks for your contribution! The author(s) have been notified to review your proposed change.

davidebeatrici commented:
Thanks for fixing the writing and merging the PR!
